Source/Dafny/Resolver.cs

```diff
@@ -9625,295 +9688,888 @@ public class ResolveTypeReturn
       }
     }
 
-    /*
-     * Convert
-     *   match xs
-     *     case Cons(y, Cons(z, zs)) => last(Cons(z, zs))
-     *     case Cons(y, Nil) => y
-     * To
-     *   match xs
-     *     case Cons(y, ys) => match ys
-     *       case Nil => y
-     *       case Cons(z, zs) => last(ys)
-     */
-    void DesugarMatchCaseStmt(MatchStmt s, DatatypeDecl dtd, List<Tuple<CasePattern<BoundVar>, BoundVar>> patterns, ICodeContext codeContext) {
-      Contract.Assert(dtd != null);
-      Dictionary<string, DatatypeCtor> ctors = datatypeCtors[dtd];
-      if (ctors == null) {
-        // there is no constructor, no need to desugar
-        return;
+
+/* Temporary information about the Match being desugared  */
+public class MatchTempInfo{
+  public IToken Tok;
+
```

Comment:
This is still in the lexical scope of the previous class, right? If so, these things should be indented.